File: chap15_muscle/exercises/muscle_compute_max_v.py
import numpy as np
from scipy.integrate import solve_ivp
from scipy.optimize import fsolve
import matplotlib.pyplot as plt


def deriv(x,n,v):
    ndot = -(1/v)*(f(x)*(1-n) - g(x)*n)   
    return ndot

def f(x):
    return np.piecewise(x,[x<=0,0<=x<=1,x>1],[0,f1*x,0])
    
def g(x):
    return np.piecewise(x,[x<=0,x>0],[g2,g1*x])

def get_load(v):
    soln = solve_ivp(lambda x,n: deriv(x,n,v),[2, -5],[n0],method='LSODA',max_step=0.005)
    return np.trapz(soln.t,soln.t*soln.y[0])

# ----------------------------------------------------------------

f1=43.3
g1=10 
g2=209
vv = [1,10,50,100,120]

# plot the n for a range of v
# Initial conditions
n0 = 0
plt.figure(1)
for i in range(0,5):
    v = vv[i]
    # You have to prevent the solver taking big steps. It wants to. It's very naughty.
    soln = solve_ivp(lambda x,n: deriv(x,n,v),[2, -3],[n0],method='LSODA',max_step=0.005)
    plt.plot(soln.t,soln.y[0])
plt.xlabel('$x')
plt.ylabel('$n(x)$')

# next, plot the force-velocity curve
p = np.empty(20)
vv = np.empty(20)
for i in range(0,20):
    v = 0.1 + i*120/20
    p[i] = get_load(v)
    vv[i] = v

plt.figure(2)
plt.plot(p,vv)
plt.xlabel('$p$')
plt.ylabel('$v$')

# Plot the maximal velocity as a function of g2
gg2 = np.empty(20)
vmax = np.empty(20)
for i in range(0,20):
    g2 = 150 + i*100/20
    gg2[i] = g2
    vmax[i] = fsolve(get_load,100)
plt.figure(3)
plt.plot(gg2,vmax)
plt.xlabel('$g_2$')
plt.ylabel('$v_{max}$')
    
# Plot the maximal velocity as a function of g1
# first reset the values
f1=43.3
g1=10 
g2=209
gg1 = np.empty(20)
vmax = np.empty(20)
for i in range(0,20):
    g1 = 5 + i*10/20
    gg1[i] = g1
    vmax[i] = fsolve(get_load,107)
plt.figure(4)
plt.plot(gg1,vmax)
plt.xlabel('$g_1$')
plt.ylabel('$v_{max}$')
plt.ylim([105,110])

# Plot the maximal velocity as a function of f1
# first reset the values
f1=43.3
g1=10 
g2=209
ff1 = np.empty(20)
vmax = np.empty(20)
for i in range(0,20):
    f1 = 20 + i*80/20
    ff1[i] = f1
    vmax[i] = fsolve(get_load,107)
plt.figure(5)
plt.plot(ff1,vmax)
plt.xlabel('$f_1$')
plt.ylabel('$v_{max}$')
plt.ylim([105,110])
    

File: chap15_muscle/exercises/muscle_compute_max_v.m

function muscle_compute_max_v

close all
clear all
clc

par.f1=43.3; 
par.g1=10;
par.g2=209;
par.h=1;
par.v = 100; % constant velocity of contraction

%%  First calculate n for various v, and plot
xspan = linspace(2,-8,1500);
n0 = 0;
figure(1)
hold on
v = [1 10 50 100 120];
for i=1:5
par.v = v(i);
[xout,nout]=ode15s(@(x,n)derivs(x,n,par),xspan,n0);
plot(xout,nout,'LineWidth',2)
end
hold off
xlim([-2 2])
xlabel('x')
ylabel('n')
exportgraphics(gcf,'../../../Math-Physiol-3rd-edition/figures/chap_15_muscle/exercises/muscle_compute_max_v_1.pdf')

%% Next, plot the force-velocity curve for the given parameters
for i = 1:20
    v(i) = 0.1 + (i-1)*120/19;
    p(i) = get_load(v(i),par);
end
figure(2)
plot(p,v,'LineWidth',2)
xlabel('force')
ylabel('velocity')
exportgraphics(gcf,'../../../Math-Physiol-3rd-edition/figures/chap_15_muscle/exercises/muscle_compute_max_v_2.pdf')

%% Finally, plot the maximal velocity as a function of each parameter. Reset the parameter values for each plot.
par.f1=43.3; 
par.g1=10;
par.g2=209;
for i = 1:20
    par.g1 = 5 + (i-1)*10/19;
    g1(i) = par.g1;
    vmax(i) = fzero(@(v)get_load(v,par),100); % The maximal velocity occurs when the load = 0
end
figure(3)
plot(g1,vmax,'LineWidth',2)
xlabel('g_1')
ylabel('v_{max}')
exportgraphics(gcf,'../../../Math-Physiol-3rd-edition/figures/chap_15_muscle/exercises/muscle_compute_max_v_3.pdf')

par.f1=43.3; 
par.g1=10;
par.g2=209;
for i = 1:20
    par.g2 = 150 + (i-1)*100/19;
    g2(i) = par.g2;
    vmax(i) = fzero(@(v)get_load(v,par),100); % The maximal velocity occurs when the load = 0
end
figure(4)
plot(g2,vmax,'LineWidth',2)
xlabel('g_2')
ylabel('v_{max}')
exportgraphics(gcf,'../../../Math-Physiol-3rd-edition/figures/chap_15_muscle/exercises/muscle_compute_max_v_4.pdf')

par.f1=43.3; 
par.g1=10;
par.g2=209;
for i = 1:20
    par.f1 = 20 + (i-1)*70/19;
    f1(i) = par.f1;
    vmax(i) = fzero(@(v)get_load(v,par),100); % The maximal velocity occurs when the load = 0
end
figure(5)
plot(f1,vmax,'LineWidth',2)
xlabel('f_1')
ylabel('v_{max}')
exportgraphics(gcf,'../../../Math-Physiol-3rd-edition/figures/chap_15_muscle/exercises/muscle_compute_max_v_5.pdf')

end

%% -----------------------------------------------
function load = get_load(v,par)
% initial conditions
xspan = linspace(2,-8,1500);
n0 = 0;
par.v = v;
[xout,nout]=ode15s(@(x,n)derivs(x,n,par),xspan,n0);
load = -trapz(xout,xout.*nout);  % Has to be minus since x is going backwards in the integration
end

%% ------------------------------------------------
% RHS of ode
function out = derivs(x,n,par)
out = (-1/par.v)*( (1-n)*f(x,par) - n*g(x,par) );
end

%% ------------
function out = f(x,par)
out = 0 + (x>0 & x<1).*(par.f1*x);
end

function out = g(x,par)
out = par.g2*(x<=0) + par.g1*x.*(x>0);
end







